obal/tito_extensions/git_source_strategy.py

```diff
@@ -0,0 +1,183 @@
+import json
+import os
+import os.path
+import re
+import shutil
+import subprocess
+from zipfile import ZipFile
+
+from tito.builder.fetch import SourceStrategy
+from tito.common import error_out, debug, run_command
+
+class GitSourceStrategy(SourceStrategy):
+    """
+    Designed to be used for nightly or pull request builds only.
+    It first copies source files from git, then copies the source file(s)
+    over the top, so they're merged (patches etc can then be stored
+    in git).
+    """
+    def fetch(self):
+        if "source_dir" not in self.builder.args:
+            raise Exception("Specify '--arg source_dir=...'")
+
+        # Copy the live spec from our starting location. Unlike most builders,
+        # we are not using a copy from a past git commit.
+        self.spec_file = os.path.join(self.builder.rpmbuild_sourcedir,
+                    '%s.spec' % self.builder.project_name)
+        shutil.copyfile(
+            os.path.join(self.builder.start_dir, '%s.spec' %
+                self.builder.project_name),
+            self.spec_file)
+
+        gitrev = self._fetch_local()
+
+        for s in os.listdir(self.builder.start_dir):
+            if os.path.exists(os.path.join(self.builder.start_dir, s)):
+                shutil.copyfile(
+                    os.path.join(self.builder.start_dir, s),
+                    os.path.join(self.builder.rpmbuild_sourcedir, os.path.basename(s)))
+        print("  %s.spec" % self.builder.project_name)
+
+        replacements = []
+        src_files = run_command("find %s -type f" %
+              os.path.join(self.builder.rpmbuild_sourcedir, 'archive')).split("\n")
+
+        def filter_archives(path):
+            base_name = os.path.basename(path)
+            return ".tar" in base_name
+
+        for i, s in enumerate(filter(filter_archives, src_files)):
+            base_name = os.path.basename(s)
+            debug("Downloaded file %s" % base_name)
+
+            dest_filepath = os.path.join(self.builder.rpmbuild_sourcedir,
+                    base_name)
+            shutil.move(s, dest_filepath)
+            self.sources.append(dest_filepath)
+
+            # Add a line to replace in the spec for each source:
+            source_regex = re.compile("^(source%s:\s*)(.+)$" % i, re.IGNORECASE)
+            new_line = "Source%s: %s\n" % (i, base_name)
+            replacements.append((source_regex, new_line))
+
+        # Replace version in spec:
+        version_regex = re.compile("^(version:\s*)(.+)$", re.IGNORECASE)
+        self.version = self._get_version()
+        print("Building version: %s" % self.version)
+        replacements.append((version_regex, "Version: %s\n" % self.version))
+        self.replace_in_spec(replacements)
+
+        rel_date = datetime.utcnow().strftime("%Y%m%d%H%M")
+        self.release = rel_date + gitrev
+        print("Building release: %s" % self.release)
+        run_command("sed -i '/^Release:/ s/%%/.%s%%/' %s" % (self.release, self.spec_file))
```

Comment:
Should this also use `self.replace_in_spec`?

Reply:
Possibly.  It may be cleaner.  This was slightly modified from the following block of code to replace the release macro instead:

https://github.com/theforeman/foreman-packaging/blob/rpm/develop/rel-eng/custom/custom.py#L88-L91